---
title: Quality Control
parent: Read Mapping
nav_order: 3
---

# Quality Control

## Generate QC Files for MultiQC

A bash script is provided to simplify the QC procedure. It generates QC
files that can be picked up by [MultiQC](https://multiqc.info/).
```bash
$ ./scripts/QC.sh -v input.vcf -o output_dir \
    /path/to/reference_assets /path/to/my_reference.fa sample_name input.bam
```
You can find zip files with supplementary QC asset files on the 
[latest release page](https://github.com/huishenlab/biscuit/releases/latest) on
GitHub. These additional files are provided for hg19, hg38, and mm10, as well as
a copy of the QC.sh file for those who downloaded the binary version from
GitHub. Asset files for other genome builds would need to be generated by the
user (see the [FAQ]({{ site.baseurl }}{% link docs/faq.md %}) page for more
details on creating these files).

## Filter Reads by Bisulfite Conversion

For some library preparations, incomplete conversions are enriched in a subset
of reads which need to be filtered. The `bsconv` subcommand transforms an input
BAM into one that contains a `ZN` tag (like
`ZN:Z:CA_R0C11,CC_R1C14,CG_R0C2,CT_R1C5`). This tag summarizes counts of
retention and conversion for four different cytosine contexts `CpA`, `CpC`,
`CpG` and `CpT`. The `-b` flag outputs the counts in a table, instead of as a
tag in the BAM file.
```bash
$ biscuit bsconv /path/to/my_reference.fa input.bam
```

For help on available flags, enter `biscuit bsconv` on the command line.

## Validate Bisulfite Conversion Label

Sometimes, the bisulfite conversion labels in a given alignment are inaccurate,
conflicting, or ambiguous. The `bsstrand` subcommand summarizes these labels,
given the number of C&#8594;T and G&#8594;A substitutions. As an option, it can
also correct inaccurate labels.
```bash
$ biscuit bsstrand /path/to/my_reference.fa input.bam
```
returns something similar to:
```
Mapped reads: 2865
Unmapped reads: 44
Corrected reads: 0 (0.00%)

Strand Distribution:
strand\BS      BSW (f)      BSC (r)
     R1 (f):   67           528
     R1 (r):   692          47
     R2 (f):   765          92
     R2 (r):   107          567


R1 mapped to converted:   114
R1 mapped to synthesized: 1220
R2 mapped to converted:   1332
R2 mapped to synthesized: 199

Confusion counts:
orig\infer      BSW (f)      BSC (r)      Conflict (c) Unknown (u)
     BSW (f):   1483         48           11           89
     BSC (r):   27           1084         6            117
Conflict (c):   0            0            0            0
 Unknown (u):   0            0            0            0
```

The inferred `YD` tag gives the following
  - f: foward/Waston strand
  - r: reverse/Crick strand
  - c: conflicting strand information
  - u: unintelligible strand source (unknown)

`YD` is inferred based on the count of `C>T` (`nCT`) and `G>A` (`nGA`)
observations in each read according to the following rules:

  - If both `nCT` and `nGA` are zero, `YD = u` and `s = min(nGA,nCT) / max(nGA,nCT)`.
  - If `nCT > nGA` and (`nGA == 0` or `s <= 0.5`), then `YD = f`.
  - If `nCT < nGA` and (`nCT == 0` or `s <= 0.5`), then `YD = r`.
  - All other scenarios give `YD = c`.

The flag `-y` appends `nCT` (`YC` tag) and `nGA` (`YG` tag) in the output BAM
file.

For more help with `bsstrand`, run `biscuit bsstrand` in the terminal.

### When `-b 1` Was Specified in Mapping

`-b 1` forces the mapping to be conducted in a stranded manner. Using `bsstrand`,
you can validate whether this enforcement is successful. For example,
```bash
$ biscuit bsstrand /path/to/my_reference.fa stranded.bam
```
returns
```
Mapped reads: 2918
Unmapped reads: 93
Corrected reads: 0 (0.00%)

Strand Distribution:
strand\BS      BSW (f)      BSC (r)
     R1 (f):   840          0
     R1 (r):   0            551
     R2 (f):   0            565
     R2 (r):   962          0


R1 mapped to converted:   1391
R1 mapped to synthesized: 0
R2 mapped to converted:   0
R2 mapped to synthesized: 1527

Confusion counts:
orig\infer      BSW (f)      BSC (r)      Conflict (c) Unknown (u)
     BSW (f):   715          916          9            162
     BSC (r):   499          500          12           105
Conflict (c):   0            0            0            0
 Unknown (u):   0            0            0            0
```
In this case, read 1 is always mapped to the converted strand and read 2 is
always mapped to the synthesized strand.

## Investigate Cytosine-Read Pair in Tabular Form

The `cinread` subcommand can print information about cytosines in various
contexts in a tab-separated format. By default the read name, whether the read
comes from read 1 or read 2, the bisulfite strand, reference base, and base
found in the read are printed. The printed information can be chosen using the
`-p` flag.

```bash
$ biscuit cinread /path/to/my_reference.fa input.bam
```

For more information about the available options, run `biscuit cinread` in your
terminal.
